nemo/collections/tts/fastspeech_modules.py

```diff
@@ -0,0 +1,373 @@
+from nemo.collections.asr.parts import AudioDataset, WaveformFeaturizer
+from nemo.collections.tts.parts import fastspeech, transformer
+from nemo.core.neural_types import AudioSignal, EmbeddedTextType, LengthsType, MaskType, MelSpectrogramType, NeuralType
+from nemo.utils.decorators import add_port_docs
+
+__all__ = ['FastSpeechDataLayer', 'FastSpeech', 'FastSpeechLoss']
+
+
+class FastSpeechDataLayer(DataLayerNM):
+    """Data Layer for Fast Speech model.
+
+    Basically, replicated behavior from AudioToText Data Layer, zipped with ground truth durations for additional loss.
+
+    Args:
+        manifest_filepath (str): Dataset parameter.
+            Path to JSON containing data.
+        durs_dir (str): Path to durations arrays directory.
+        labels (list): Dataset parameter.
+            List of characters that can be output by the ASR model.
+            For Jasper, this is the 28 character set {a-z '}. The CTC blank
+            symbol is automatically added later for models using ctc.
+        batch_size (int): batch size
+        sample_rate (int): Target sampling rate for data. Audio files will be
+            resampled to sample_rate if it is not already.
+            Defaults to 16000.
+        int_values (bool): Bool indicating whether the audio file is saved as
+            int data or float data.
+            Defaults to False.
+        eos_id (id): Dataset parameter.
+            End of string symbol id used for seq2seq models.
+            Defaults to None.
+        min_duration (float): Dataset parameter.
+            All training files which have a duration less than min_duration
+            are dropped. Note: Duration is read from the manifest JSON.
+            Defaults to 0.1.
+        max_duration (float): Dataset parameter.
+            All training files which have a duration more than max_duration
+            are dropped. Note: Duration is read from the manifest JSON.
+            Defaults to None.
+        normalize_transcripts (bool): Dataset parameter.
+            Whether to use automatic text cleaning.
+            It is highly recommended to manually clean text for best results.
+            Defaults to True.
+        trim_silence (bool): Whether to use trim silence from beginning and end
+            of audio signal using librosa.effects.trim().
+            Defaults to False.
+        load_audio (bool): Dataset parameter.
+            Controls whether the dataloader loads the audio signal and
+            transcript or just the transcript.
+            Defaults to True.
+        drop_last (bool): See PyTorch DataLoader.
+            Defaults to False.
+        shuffle (bool): See PyTorch DataLoader.
+            Defaults to True.
+        num_workers (int): See PyTorch DataLoader.
+            Defaults to 0.
+        perturb_config (dict): Currently disabled.
+
+    """
+
+    @property
+    @add_port_docs
+    def output_ports(self):
+        """Returns definitions of module output ports."""
```

Comment:
@okuchaiev Do we need this docstring?

Reply:
You mean "Returns definitions of module output ports."? Pretty much every NeMo module has it right now.

Reply:
Yes but it's a useless docstring so why bother adding it.

Reply:
This is not blocking your PR, but a general question posed at @okuchaiev